pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,208 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package telemetry
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/elastic/cloud-on-k8s/pkg/about"
+	apmv1 "github.com/elastic/cloud-on-k8s/pkg/apis/apm/v1"
+	beatv1beta1 "github.com/elastic/cloud-on-k8s/pkg/apis/beat/v1beta1"
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	entv1beta1 "github.com/elastic/cloud-on-k8s/pkg/apis/enterprisesearch/v1beta1"
+	kbv1 "github.com/elastic/cloud-on-k8s/pkg/apis/kibana/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/kibana"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	"github.com/ghodss/yaml"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	updateInterval = 10 * time.Second
+	resourceCount  = "resource_count"
+	podCount       = "pod_count"
+)
+
+var log = logf.Log.WithName("usage")
+
+type getUsageFn func(k8s.Client, []string) (string, interface{}, error)
+
+// ECK is a helper struct to marshal telemetry information.
+type ECK struct {
+	ECK   about.OperatorInfo     `json:"eck"`
+	Usage map[string]interface{} `json:"eck_usage"`
+}
+
+func NewReporter(info about.OperatorInfo, client client.Client, managedNamespaces []string) Reporter {
+	if len(managedNamespaces) == 0 {
+		// treat no managed namespaces as managing all namespaces, ie. set empty string for namespace filtering
+		managedNamespaces = append(managedNamespaces, "")
+	}
+
+	return Reporter{
+		operatorInfo:      info,
+		client:            k8s.WrapClient(client),
+		managedNamespaces: managedNamespaces,
+	}
+}
+
+type Reporter struct {
+	operatorInfo      about.OperatorInfo
+	client            k8s.Client
+	managedNamespaces []string
+}
+
+func (r *Reporter) Start() {
+	ticker := time.NewTicker(updateInterval)
+	for range ticker.C {
+		usage, err := r.getResourceUsage()
+		if err != nil {
+			log.Error(err, "failed to get resource usage")
+			continue
+		}
+
+		telemetryBytes, err := yaml.Marshal(ECK{ECK: r.operatorInfo, Usage: usage})
+		if err != nil {
+			log.Error(err, "failed to marshal telemetry data")
+		}
+
+		for _, ns := range r.managedNamespaces {
+			var kibanaList kbv1.KibanaList
+			if err := r.client.List(&kibanaList, client.InNamespace(ns)); err != nil {
+				log.Error(err, "failed to list kibanas")
+				continue
+			}
+			for _, kb := range kibanaList.Items {
+				var secret corev1.Secret
+				nsName := types.NamespacedName{Namespace: kb.Namespace, Name: kibana.SecretName(kb)}
+				if err := r.client.Get(nsName, &secret); err != nil {
+					log.Error(err, "failed to get kibana secret")
+					continue
+				}
+
+				secret.Data[kibana.TelemetryFilename] = telemetryBytes
```

Comment:
Can we be smart here and only update if there is a change in data. I am bit worried about the number of updates we will be pushing to the API server in larger ECK operator installations with many Kibanas.